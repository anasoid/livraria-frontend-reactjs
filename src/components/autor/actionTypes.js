export const GET_AUTORES = 'GET_AUTORES';
export const ADD_AUTOR = 'ADD_AUTOR';
export const REMOVE_AUTOR = 'REMOVE_AUTOR';