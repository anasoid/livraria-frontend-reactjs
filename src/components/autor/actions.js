import Helper from './helper';
import {GET_AUTORES} from './types';

export const getAutoresList = () => {
  const helper = new Helper();
  return dispatch => {
    return helper.getAutoresList()
      .then((data) => {
        return dispatch({type: GET_AUTORES, payload: data});
      }).catch();
  };
};

