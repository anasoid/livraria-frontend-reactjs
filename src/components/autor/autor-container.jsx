import React from 'react';
import AutorTable from './autor-table';

const AutorContainer = (props) => {
  return(
    <div>
      <h1>Novo Autor</h1>
      <form>
        <label htmlFor={'autorNome'}>Nome: </label>
        <input id={'autorNome'} name={'nome'}/>
        <br/>
        <label htmlFor={'autorEmail'}>Email: </label>
        <input id={'autorEmail'} name={'email'}/>
        <br/>
        <button id={'btnGravarAutor'} type={'submit'}>Gravar Autor</button>
      </form>
      <br/>
      <AutorTable/>
    </div>
  );
};

export default AutorContainer;