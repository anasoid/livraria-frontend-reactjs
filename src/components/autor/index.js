import React from 'react';
import AutorContainer from './autor-container';

const Autor = () => {
  return (
    <AutorContainer/>
  );
};

export default Autor;