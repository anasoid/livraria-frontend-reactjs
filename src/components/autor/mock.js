export const autores = {
  "autores": [
    {'nome': 'fulano', 'email': 'fulano@fulano.com'},
    {'nome': 'ciclano', 'email': 'ciclano@ciclano.com'}
  ]
};