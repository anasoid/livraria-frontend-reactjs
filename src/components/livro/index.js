import React from 'react';
import LivroContainer from './livro-container';

const Livro = () => {
  return (
    <LivroContainer/>
  );
}

export default Livro;