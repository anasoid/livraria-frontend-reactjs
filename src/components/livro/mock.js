export const livros = {
  livros: [
    {titulo: 'reactjs', isbn: '12345678', preco: 20.0, dtLancamento: '11/05/2018', autores: [{nome: 'fulano', email: 'fulano@email.com'}]},
    {titulo: 'react native', isbn: '87654321', preco: 15.0, dtLancamento: '20/08/2018', autores: [{nome: 'ciclano', email: 'ciclano@email.com'}]}
  ]
}