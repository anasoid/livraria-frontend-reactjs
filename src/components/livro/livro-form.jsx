import React, {Component} from 'react';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import {getAutoresList} from '../autor/actions';

export class LivroForm extends Component{

  constructor(props) {
    super(props);
    this.state = {
      selectAutor: '',
      autoresSelected: []
    }
  }

  componentDidMount() {
    this.props.getAutoresList();
  }

  render() {
    return (
      <div>
        <form>
          <label htmlFor={'livroTitulo'}>Titulo: </label>
          <input id={'livroTitulo'} name={'titulo'}/>
          <br/>
          <label htmlFor={'livroIsbn'}>ISBN: </label>
          <input id={'livroIsbn'} name={'isbn'}/>
          <br/>
          <label htmlFor={'livroPreco'}>Preço: </label>
          <input id={'livroPreco'} name={'preco'}/>
          <br/>
          <label htmlFor={'livroDataLancamento'}>Data de Lançamento: </label>
          <input id={'livroDataLancamento'} name={'dataLancamento'}/>
          <br/>
          <label htmlFor={'livroAutores'}>Selecione o Autor: </label>
          <select id={'livroAutores'}
            name={'selectAutor'}
            value={this.state.selectAutor}
            onChange={this.handleChange.bind(this)}>
            <option value={''}>Selecione</option>
            {
              this.props.autores.map((autor, index) => {
                return (
                  <option key={index} value={index}>{autor.nome}</option>
                );
              })
            }
          </select>
          <button id={'btnAddAutor'} type={'button'} onClick={this.addAutorToList.bind(this)}>Adicionar Autor</button>
          <br/>
          <label>Autores: </label>
          <ul>
            {
              this.state.autoresSelected.map((autor, index) => {
                return (
                  <li key={index}>{autor.nome}</li>
                );
              })
            }
          </ul>
          <br/>
          <button id={'btnGravar'} type={'submit'} disabled={true}>Gravar Livro</button>
        </form>
      </div>
    );
  }

  addAutorToList() {
    let autoresSelected = [...this.state.autoresSelected].concat(this.props.autores[this.state.selectAutor]);
    this.setState({
      autoresSelected,
      selectAutor: ''
    });
  }

  handleChange(event) {
    this.setState({
      selectAutor: event.target.value
    });
  }
}

const mapStateToProps = state => ({
  autores: state.autor.autores
});

const mapDispatchToProps = dispatch => bindActionCreators({
  getAutoresList
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(LivroForm);