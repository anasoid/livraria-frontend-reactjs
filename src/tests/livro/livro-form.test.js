import React from 'react';
import {mount} from 'enzyme/build/index';
import renderer from 'react-test-renderer';
import {LivroForm} from '../../components/livro/livro-form';
import {autores} from '../../components/autor/mock';

const getAutoresListMock = jest.fn();
const component = (
  <LivroForm
    autores={autores.autores}
    getAutoresList={getAutoresListMock}
  />
);

describe('test livroForm component', () => {

  let wrapper;

  beforeEach(() => {
    wrapper = mount(component);
  });

  it('should renders without crashing', () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
    expect(getAutoresListMock).toBeCalled();
  });

  it('should be input titulo do livro', () => {
    const titulo = wrapper.find('#livroTitulo');
    expect(titulo).toMatchSelector('input');
    expect(titulo).toHaveProp('name', 'titulo');
  });

  it('should be input isbn do livro', () => {
    const isbn = wrapper.find('#livroIsbn');
    expect(isbn).toMatchSelector('input');
    expect(isbn).toHaveProp('name', 'isbn');
  });

  it('should be input Preço', () => {
    const preco = wrapper.find('#livroPreco');
    expect(preco).toMatchSelector('input');
    expect(preco).toHaveProp('name', 'preco');
  });

  it('should be input Data Lançamento', () => {
    const dataLancamento = wrapper.find('#livroDataLancamento');
    expect(dataLancamento).toMatchSelector('input');
    expect(dataLancamento).toHaveProp('name', 'dataLancamento');
  });

  it('should be multiselect Autor', () => {
    const autores = wrapper.find('#livroAutores');
    expect(autores).toMatchSelector('select');
    expect(autores).toHaveProp('name', 'selectAutor');
  });

  it('should be button adicionar autor', () => {
    const btnAutor = wrapper.find('#btnAddAutor');
    expect(btnAutor).toMatchSelector('button');
    expect(btnAutor).toHaveProp('type', 'button');
    expect(btnAutor).toHaveText('Adicionar Autor');
  });

  it('should be form to submit livro', () => {
    expect(wrapper.find('form')).toExist();
  });

  it('should be button submit Gravar Livro', () => {
    const button = wrapper.find('#btnGravar');
    expect(button).toMatchSelector('button');
    expect(button).toHaveProp('type', 'submit');
    expect(button).toHaveText('Gravar Livro');
  });

  it('should disabled button Gravar Livro when inputs is empty', () => {
    const button = wrapper.find('#btnGravar').at(0);
    expect(button).toHaveProp('disabled', true);
  });

  it('should select autor when change select value', () => {
    const select = wrapper.find('#livroAutores').at(0);
    select.simulate('change', {target: { value : 0}});
    expect(wrapper.state().selectAutor).toEqual(0);
  });

  it('should add autor do list when click Adicionar Autor button', async () => {
    await wrapper.setState({selectAutor: 0});
    const btnAutor = wrapper.find('#btnAddAutor').at(0);
    btnAutor.simulate('click');

    expect(wrapper.state().autoresSelected.length).toEqual(1);
    expect(wrapper.state().selectAutor).toEqual('');
    expect(wrapper.state().autoresSelected).toContain(autores.autores[0]);
  });

});