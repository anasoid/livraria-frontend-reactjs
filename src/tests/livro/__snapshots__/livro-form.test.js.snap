// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test livroForm component should renders without crashing 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="livroTitulo"
    >
      Titulo: 
    </label>
    <input
      id="livroTitulo"
      name="titulo"
      onChange={[Function]}
      value=""
    />
    <br />
    <label
      htmlFor="livroIsbn"
    >
      ISBN: 
    </label>
    <input
      id="livroIsbn"
      name="isbn"
      onChange={[Function]}
      value=""
    />
    <br />
    <label
      htmlFor="livroPreco"
    >
      Preço: 
    </label>
    <input
      id="livroPreco"
      name="preco"
      onChange={[Function]}
      value={0}
    />
    <br />
    <label
      htmlFor="livroDataLancamento"
    >
      Data de Lançamento: 
    </label>
    <input
      id="livroDataLancamento"
      name="dataLancamento"
      onChange={[Function]}
      value=""
    />
    <br />
    <label
      htmlFor="livroAutores"
    >
      Selecione o Autor: 
    </label>
    <select
      id="livroAutores"
      name="selectAutor"
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      >
        Selecione
      </option>
      <option
        value={0}
      >
        fulano
      </option>
      <option
        value={1}
      >
        ciclano
      </option>
    </select>
    <button
      id="btnAddAutor"
      onClick={[Function]}
      type="button"
    >
      Adicionar Autor
    </button>
    <br />
    <label>
      Autores: 
    </label>
    <ul />
    <br />
    <button
      disabled={true}
      id="btnGravar"
      type="submit"
    >
      Gravar Livro
    </button>
  </form>
</div>
`;
